daemon/cluster/cluster.go

```diff
@@ -871,6 +872,36 @@ func (c *Cluster) GetServices(options apitypes.ServiceListOptions) ([]types.Serv
 	return services, nil
 }
 
+// imageWithDigestString takes an image such as name or name:tag
+// and returns the image pinned to a digest, such as name@sha256:34234...
+func (c *Cluster) imageWithDigestString(ctx context.Context, image, encodedAuth string) (string, error) {
```

Comment:
That is probably a first step. It is hard to tell, since it is not actually being used. I have no clue why we would bury an unencoded value so deeply in the call stack. The auth should probably be decoded as soon as the request is received. Then, I would put it in the context and access as needed, without passing it down the entire call chain.